### DISCUSSION TITLE
How is current span shared across the context?

### DISCUSSION THREAD

**Discussion by agentshowers** [open-telemetry/opentelemetry-ruby#885] [Q&A]
Hello everyone.

I'm still very new at open telemetry, we're currently just playing around with it on our production setup. We have some other instrumentation in place and were trying to re-use the same hooks for Open Telemetry. To do it, I was expecting that once a span is started (via `start_root_span`) would update the current span. However, if I do something like this, the IDs are different:

```
tracer = OpenTelemetry.tracer_provider.tracer('my-app', '0.1.0')

span = tracer.start_root_span('my-span')
p "Span ID: #{parent_span.context.span_id}"

current_span = OpenTelemetry::Trace.current_span
p "Current ID: #{current_span.context.span_id}"
```

`Span ID` and `Current ID` are two different IDs while I expected them to be same. Eventually what would want to do is something like this:

```
TRACER = OpenTelemetry.tracer_provider.tracer('my-app', '0.1.0')

def start_span
  if current_span_exists
    TRACER.start_span('some-child-span', with_parent: current_context)
  else
    TRACER.start_root_span('some-root-span')
  end
end

def end_span
  OpenTelemetry::Trace.current_span.finish
end
```

These two methods would be called by the hooks in my code and would either start a root span or a child span with the current span as the parent. All of this is inside the same request and thread. Is there a way to achieve this that I'm missing? All the examples I've seen so far use the `in_span` way, haven't figured out how to properly use the `start_span` / `start_root_span` methods

**Comment by arielvalentin** [comment]
@agentshowers The value of using `in_span` as a convenience method is that it manages the current `Span` using the `Context` object. If you start your own spans manually you will have to keep track of them yourself as well as set the `Context` state to take advantage of auto-instrumentation. 

Here is how `in_span` manages to set the span as part of the current context:
https://github.com/open-telemetry/opentelemetry-ruby/blob/f08c802f726b792c029d4f50a878dba69398ff56/api/lib/opentelemetry/trace/tracer.rb#L29

Is using convenience methods like `in_span` something you will not be able to adopt in your application?

**Reply by agentshowers** [reply]
We will be able to do it, and probably want to. But right now we were just trying to do some monkey patching on some existing instrumentation (ddtrace) so we can send all of our existing traces using both systems. Since this is just a quick 'n dirty patch to do some testing, I'll play around with the `Context` for a while and if not I'll do it the proper way.
